lane.py

```diff
@@ -2138,99 +1664,325 @@ def _featured_works_with_lanes(
+        )
 
-        :param target_size: Try to get approximately this many
-        items. There may be more or less; this controls the size of
-        the window and the LIMIT on the query.
+    @classmethod
+    def _modify_loading(cls, qu):
+        """Optimize a query for use in generating OPDS feeds, by modifying
+        which related objects get pulled from the database.
         """
-        lane_query = self.works_from_database(_db, facets=facets)
+        # Avoid eager loading of objects that are already being loaded.
+        qu = qu.options(
+            contains_eager(Work.presentation_edition),
+            contains_eager(Work.license_pools),
+        )
+        license_pool_name = 'license_pools'
+
+        # Load some objects that wouldn't normally be loaded, but
+        # which are necessary when generating OPDS feeds.
+
+        # TODO: Strictly speaking, these joinedload calls are
+        # only needed by the circulation manager. This code could
+        # be moved to circulation and everyone else who uses this
+        # would be a little faster. (But right now there is no one
+        # else who uses this.)
+        qu = qu.options(
+            # These speed up the process of generating acquisition links.
+            joinedload(license_pool_name, "delivery_mechanisms"),
+            joinedload(license_pool_name, "delivery_mechanisms", "delivery_mechanism"),
 
-        # Make sure this query finds a number of works proportinal
-        # to the expected size of the lane.
-        lane_query = self._restrict_query_to_window(
-            lane_query, target_size, facets
+            # These speed up the process of generating the open-access link
+            # for open-access works.
+            joinedload(license_pool_name, "delivery_mechanisms", "resource"),
+            joinedload(license_pool_name, "delivery_mechanisms", "resource", "representation"),
         )
+        return qu
 
-        lane_query = lane_query.order_by(
-            text("quality_tier desc"), mw.random.desc()
+    @classmethod
+    def _defer_unused_fields(cls, query):
+        """Some applications use the simple OPDS entry and some
+        applications use the verbose. Whichever one we don't need,
+        we can stop from even being sent over from the
+        database.
+        """
+        if Configuration.DEFAULT_OPDS_FORMAT == "simple_opds_entry":
+            return query.options(defer(Work.verbose_opds_entry))
+        else:
+            return query.options(defer(Work.simple_opds_entry))
+
+
+class DatabaseBackedWorkList(WorkList):
+    """A WorkList that can get its works from the database in addition to
+    (or possibly instead of) the search index.
+
+    Even when works _are_ obtained through the search index, a
+    DatabaseBackedWorkList is then created to look up the Work objects
+    for use in an OPDS feed.
+    """
+
+    def works_from_database(self, _db, facets=None, pagination=None, **kwargs):
```

Comment:
Here's the code that I removed from `WorkList`. Since it no longer uses the materialized view, it should only be used when performance is not important or when there's no way to get what you want using the normal `works()` method.